engines/ic3sa.cpp

```diff
@@ -0,0 +1,875 @@
+        Term cv = ts_.curr(ntv);
+        auto it = state_updates.find(cv);
+        if (it != state_updates.end()) {
+          assert(ts_.only_curr(it->second));
+          get_free_symbolic_consts(it->second, projection_set_);
+        }
+      }
+    }
+  }
+}
+
+// IC3SA specific methods
+
+EquivalenceClasses IC3SA::get_equivalence_classes_from_model(
+    const UnorderedTermSet & to_keep) const
+{
+  // assumes the solver state is sat
+  EquivalenceClasses ec;
+  for (const auto & elem : term_abstraction_) {
+    const Sort & sort = elem.first;
+    const UnorderedTermSet & terms = elem.second;
+
+    // TODO figure out if a DisjointSet is a better data structure
+    //      will need to keep track of all terms in each partition though
+
+    ec[sort] = std::unordered_map<smt::Term, smt::UnorderedTermSet>();
+    std::unordered_map<smt::Term, smt::UnorderedTermSet> & m = ec.at(sort);
+    for (const auto & t : terms) {
+      if (in_projection(t, to_keep)) {
+        Term val = solver_->get_value(t);
+        m[val].insert(t);
+      }
+    }
+  }
+  return ec;
+}
+
+void IC3SA::construct_partition(const EquivalenceClasses & ec,
+                                UnorderedTermSet & out_cube) const
+{
+  // now add to the cube expressing this partition
+  for (const auto & sortelem : ec) {
+    const Sort & sort = sortelem.first;
+
+    // TODO: play around with heuristics for the representative
+    //       to add disequalities over
+    //       e.g. we're not adding all possible disequalities,
+    //       just choosing a representative from each equivalence
+    //       class and adding a disequality to encode the distinctness
+
+    //       current priority is: symbol > generic term > value
+
+    // representatives of the different classes of this sort
+    TermVec representatives;
+    Term lit;
+    for (const auto & elem : sortelem.second) {
+      const Term & val = elem.first;
+      assert(val->is_value());
+
+      const UnorderedTermSet & terms = elem.second;
+      UnorderedTermSet::const_iterator end = terms.cend();
+      UnorderedTermSet::const_iterator it = terms.cbegin();
+      Term last = *it;
+      it++;
+
+      Term repr = last;
+      bool found_repr = false;
+      bool repr_val = repr->is_value();
+
+      while (it != end) {
+        const Term & term = *(it++);
+        assert(last->get_sort() == term->get_sort());
+        lit = solver_->make_term(Equal, last, term);
+        if (!lit->is_value()) {
+          // only add if not trivially true
+          out_cube.insert(lit);
+          assert(solver_->get_value(lit) == solver_true_);
+        }
+        last = term;
+
+        // TODO: see if a DisjointSet would make this easier
+        // update representative for this class
+        if (!found_repr) {
+          if (term->is_symbolic_const()) {
+            repr = term;
+            repr_val = false;
+            found_repr = true;
+          } else if (!term->is_value() && repr_val) {
```

Comment:
is the else case possible?

soltelem.second contains variables, right?

Reply:
It actually contains interpreted constants also. The term abstraction includes everything. This is actually important cause you might need to learn that `x = 2` or something. It also guarantees completeness I think.